File: test-arm-simulator.js
#!/usr/bin/env node

/**
 * BACA Robotic Arm Simulator
 *
 * This script simulates a robotic arm sending diagnostic data to the dashboard.
 * Use this to test the real-time data flow and verify your dashboard is working.
 *
 * Usage:
 *   node test-arm-simulator.js <arm_id>
 *
 * Example:
 *   node test-arm-simulator.js 3e77b0e6-e8c0-4fbb-bf74-5da012218240
 */

const SUPABASE_URL = 'https://tiwmxizhofixjuoydnvg.supabase.co';
const SUPABASE_KEY = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6InRpd214aXpob2ZpeGp1b3lkbnZnIiwicm9sZSI6ImFub24iLCJpYXQiOjE3NjM3NTQ1MzAsImV4cCI6MjA3OTMzMDUzMH0.G6BuNYTyfgEt4VXxWIfyoORztFYqESgZxxuGUuNoBVs';

const armId = process.argv[2];

if (!armId) {
  console.error('Error: Please provide an arm ID');
  console.log('Usage: node test-arm-simulator.js <arm_id>');
  console.log('\nTo get an arm ID:');
  console.log('1. Open the dashboard');
  console.log('2. Add a new robotic arm');
  console.log('3. Check the database or browser console for the arm ID');
  process.exit(1);
}

let temperature = 40;
let powerConsumption = 280;
let cpuUsage = 45;
let memoryUsage = 50;
let operationHours = 1000;
let errorCount = 0;

function randomChange(value, min, max, changeRate = 2) {
  const change = (Math.random() - 0.5) * changeRate;
  const newValue = value + change;
  return Math.max(min, Math.min(max, newValue));
}

function generateJointData() {
  return {
    j1: Math.random() * 180 - 90,
    j2: Math.random() * 180 - 90,
    j3: Math.random() * 180 - 90,
    j4: Math.random() * 360 - 180,
    j5: Math.random() * 180 - 90,
    j6: Math.random() * 360 - 180
  };
}

async function sendMetrics() {
  temperature = randomChange(temperature, 30, 55, 1);
  powerConsumption = randomChange(powerConsumption, 200, 350, 5);
  cpuUsage = randomChange(cpuUsage, 20, 90, 3);
  memoryUsage = randomChange(memoryUsage, 30, 85, 2);
  operationHours += 0.0014;

  const metrics = {
    arm_id: armId,
    temperature: Number(temperature.toFixed(1)),
    power_consumption: Number(powerConsumption.toFixed(1)),
    cpu_usage: Number(cpuUsage.toFixed(1)),
    memory_usage: Number(memoryUsage.toFixed(1)),
    error_count: errorCount,
    operation_hours: Number(operationHours.toFixed(1)),
    joint_positions: generateJointData(),
    joint_velocities: generateJointData(),
    joint_torques: {
      j1: Math.random() * 60 + 20,
      j2: Math.random() * 60 + 20,
      j3: Math.random() * 60 + 20,
      j4: Math.random() * 60 + 20,
      j5: Math.random() * 60 + 20,
      j6: Math.random() * 60 + 20
    }
  };

  try {
    const response = await fetch(`${SUPABASE_URL}/rest/v1/diagnostic_metrics`, {
      method: 'POST',
      headers: {
        'apikey': SUPABASE_KEY,
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(metrics)
    });

    if (response.ok) {
      console.log(`✓ Metrics sent - Temp: ${metrics.temperature}°C, Power: ${metrics.power_consumption}W, CPU: ${metrics.cpu_usage}%`);

      if (temperature > 50) {
        await sendAlert('warning', 'mechanical', `Temperature elevated: ${temperature.toFixed(1)}°C`);
      }
      if (powerConsumption > 320) {
        await sendAlert('warning', 'electrical', `High power consumption: ${powerConsumption.toFixed(1)}W`);
      }
    } else {
      console.error('✗ Failed to send metrics:', response.status, await response.text());
    }
  } catch (error) {
    console.error('✗ Error sending metrics:', error.message);
  }
}

async function sendAlert(severity, category, message) {
  const alert = {
    arm_id: armId,
    severity: severity,
    category: category,
    message: message,
    acknowledged: false,
    resolved: false
  };

  try {
    const response = await fetch(`${SUPABASE_URL}/rest/v1/system_alerts`, {
      method: 'POST',
      headers: {
        'apikey': SUPABASE_KEY,
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(alert)
    });

    if (response.ok) {
      console.log(`⚠ Alert sent: ${message}`);
    }
  } catch (error) {
    console.error('✗ Error sending alert:', error.message);
  }
}

async function updateStatus(status) {
  try {
    const response = await fetch(`${SUPABASE_URL}/rest/v1/robotic_arms?id=eq.${armId}`, {
      method: 'PATCH',
      headers: {
        'apikey': SUPABASE_KEY,
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({ status })
    });

    if (response.ok) {
      console.log(`✓ Status updated to: ${status}`);
    }
  } catch (error) {
    console.error('✗ Error updating status:', error.message);
  }
}

console.log('BACA Robotic Arm Simulator Started');
console.log(`Arm ID: ${armId}`);
console.log('Sending metrics every 5 seconds...');
console.log('Press Ctrl+C to stop\n');

updateStatus('online');

const interval = setInterval(sendMetrics, 5000);
sendMetrics();

process.on('SIGINT', async () => {
  console.log('\n\nShutting down...');
  clearInterval(interval);
  await updateStatus('offline');
  console.log('Goodbye!');
  process.exit(0);
});

File: test-arm-simulator.py
#!/usr/bin/env python3

"""
BACA Robotic Arm Simulator

This script simulates a robotic arm sending diagnostic data to the dashboard.
Use this to test the real-time data flow and verify your dashboard is working.

Usage:
    python test-arm-simulator.py <arm_id>

Example:
    python test-arm-simulator.py 3e77b0e6-e8c0-4fbb-bf74-5da012218240
"""

import sys
import time
import json
import random
import signal
import requests

SUPABASE_URL = 'https://tiwmxizhofixjuoydnvg.supabase.co'
SUPABASE_KEY = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6InRpd214aXpob2ZpeGp1b3lkbnZnIiwicm9sZSI6ImFub24iLCJpYXQiOjE3NjM3NTQ1MzAsImV4cCI6MjA3OTMzMDUzMH0.G6BuNYTyfgEt4VXxWIfyoORztFYqESgZxxuGUuNoBVs'

headers = {
    'apikey': SUPABASE_KEY,
    'Content-Type': 'application/json'
}

class RoboticArmSimulator:
    def __init__(self, arm_id):
        self.arm_id = arm_id
        self.temperature = 40.0
        self.power_consumption = 280.0
        self.cpu_usage = 45.0
        self.memory_usage = 50.0
        self.operation_hours = 1000.0
        self.error_count = 0
        self.running = True

    def random_change(self, value, min_val, max_val, change_rate=2):
        change = (random.random() - 0.5) * change_rate
        new_value = value + change
        return max(min_val, min(max_val, new_value))

    def generate_joint_data(self):
        return {
            'j1': random.uniform(-90, 90),
            'j2': random.uniform(-90, 90),
            'j3': random.uniform(-90, 90),
            'j4': random.uniform(-180, 180),
            'j5': random.uniform(-90, 90),
            'j6': random.uniform(-180, 180)
        }

    def send_metrics(self):
        self.temperature = self.random_change(self.temperature, 30, 55, 1)
        self.power_consumption = self.random_change(self.power_consumption, 200, 350, 5)
        self.cpu_usage = self.random_change(self.cpu_usage, 20, 90, 3)
        self.memory_usage = self.random_change(self.memory_usage, 30, 85, 2)
        self.operation_hours += 0.0014

        metrics = {
            'arm_id': self.arm_id,
            'temperature': round(self.temperature, 1),
            'power_consumption': round(self.power_consumption, 1),
            'cpu_usage': round(self.cpu_usage, 1),
            'memory_usage': round(self.memory_usage, 1),
            'error_count': self.error_count,
            'operation_hours': round(self.operation_hours, 1),
            'joint_positions': self.generate_joint_data(),
            'joint_velocities': self.generate_joint_data(),
            'joint_torques': {
                'j1': random.uniform(20, 80),
                'j2': random.uniform(20, 80),
                'j3': random.uniform(20, 80),
                'j4': random.uniform(20, 80),
                'j5': random.uniform(20, 80),
                'j6': random.uniform(20, 80)
            }
        }

        try:
            response = requests.post(
                f'{SUPABASE_URL}/rest/v1/diagnostic_metrics',
                headers=headers,
                json=metrics
            )

            if response.status_code == 201:
                print(f"✓ Metrics sent - Temp: {metrics['temperature']}°C, "
                      f"Power: {metrics['power_consumption']}W, "
                      f"CPU: {metrics['cpu_usage']}%")

                if self.temperature > 50:
                    self.send_alert('warning', 'mechanical',
                                  f'Temperature elevated: {self.temperature:.1f}°C')
                if self.power_consumption > 320:
                    self.send_alert('warning', 'electrical',
                                  f'High power consumption: {self.power_consumption:.1f}W')
            else:
                print(f'✗ Failed to send metrics: {response.status_code}')
        except Exception as e:
            print(f'✗ Error sending metrics: {e}')

    def send_alert(self, severity, category, message):
        alert = {
            'arm_id': self.arm_id,
            'severity': severity,
            'category': category,
            'message': message,
            'acknowledged': False,
            'resolved': False
        }

        try:
            response = requests.post(
                f'{SUPABASE_URL}/rest/v1/system_alerts',
                headers=headers,
                json=alert
            )

            if response.status_code == 201:
                print(f'⚠ Alert sent: {message}')
        except Exception as e:
            print(f'✗ Error sending alert: {e}')

    def update_status(self, status):
        try:
            response = requests.patch(
                f'{SUPABASE_URL}/rest/v1/robotic_arms?id=eq.{self.arm_id}',
                headers=headers,
                json={'status': status}
            )

            if response.status_code == 200:
                print(f'✓ Status updated to: {status}')
        except Exception as e:
            print(f'✗ Error updating status: {e}')

    def run(self):
        print('BACA Robotic Arm Simulator Started')
        print(f'Arm ID: {self.arm_id}')
        print('Sending metrics every 5 seconds...')
        print('Press Ctrl+C to stop\n')

        self.update_status('online')

        def signal_handler(sig, frame):
            print('\n\nShutting down...')
            self.running = False
            self.update_status('offline')
            print('Goodbye!')
            sys.exit(0)

        signal.signal(signal.SIGINT, signal_handler)

        self.send_metrics()

        while self.running:
            time.sleep(5)
            self.send_metrics()


if __name__ == '__main__':
    if len(sys.argv) < 2:
        print('Error: Please provide an arm ID')
        print('Usage: python test-arm-simulator.py <arm_id>')
        print('\nTo get an arm ID:')
        print('1. Open the dashboard')
        print('2. Add a new robotic arm')
        print('3. Check the database or browser console for the arm ID')
        sys.exit(1)

    arm_id = sys.argv[1]
    simulator = RoboticArmSimulator(arm_id)
    simulator.run()
